### DISCUSSION TITLE
How are demos recorded into a file?

### DISCUSSION THREAD

**Discussion by PK268** [NeKzor/SourceAutoRecord#52] [Q&A] [answered]
Hey! I've been using SAR for a long time and it only hit me now. How are the demos recorded and formatted into a file? What I mean is, when you're recording gameplay, are demos recorded frame-by-frame or by motion vectors, or something else. The reason why it is such a mystery to me is because it seems like recording something like gameplay would be such a storage-intensive process and I don't know how demos are so lightweight. Also I was wondering that if it's not something source specific. If I could try coding a similar feature for other game engines or games in specific. Thanks in advance!

**Comment by nulano** [comment, marked as answer]
I do not know the exact details as it's been a while since I've looked into it, but in essence:

Demos contain per-tick camera coordinates (Source games always run a server, even in single player, with about 60 ticks per second, exact number depends on game IIRC; also `sv_alternateticks` affects this).

Smooth camera movement at high FPS is achieved by interpolating the camera position between server ticks, and also computing the orientation on the client to react to mouse movement instantly - fast mouse movements are removed from the demo recording.

A demo file also contains all of the network traffic from the server, so that replaying a demo is as simple as "connecting to it" as if it were a server. This is what makes demos so lightweight - network code is already optimized to use as little bandwidth as possible to minimize latency, recording a demo just writes that to a file.

Demos do not contain the world geometry, that is a separate file either loaded from the game directory, or in case of e.g. CS:GO downloaded before connecting to the server. The static world geometry cannot change, e.g. moveable platforms are entities with their center point coordinates updated with a network event when they are moved.

**Reply by PK268** [reply]
Thanks! So really its just base data that's interpolated to make playback smooth. That makes sense, i guess it'll be hard to implement demos into other things without knowing how the server communicates to the client. Demos are ingenius in capturing server data however!
